merkle/smt/hstar3.go

```diff
@@ -0,0 +1,183 @@
+		for id, d := upd.ID, h.depth; d > top; d-- {
+			pref := id.Prefix(d)
+			if _, ok := ids[pref]; ok {
+				// Delete the prefix node because its original hash does not contribute
+				// to the updates, so should not be read.
+				delete(ids, pref)
+				// All the upper siblings have been added already, so skip them.
+				break
+			}
+			sib := pref.Sibling()
+			ids[sib] = nil
+		}
+	}
+	return ids
+}
+
+// Update applies the given updates to a sparse Merkle tree. Requires a
+// previous Prepare invocation on the same updates slice. Returns an error if
+// any of the NodeStorage.Get calls does so, for example if a node is missing.
+//
+// Returns the slice of updates at the top level of the sparse Merkle tree
+// induced by the provided lower level updates. Typically it will contain only
+// one item for the root hash of a tile or a (sub)tree, but the caller may
+// arrange multiple subtrees in one call, in which case the corresponding
+// returned top-level updates will be sorted lexicographically by node ID.
+//
+// Note that Update invocations can be chained. For example, a bunch of HStar3
+// instances at depth 256 can return updates for depth 8 (in parallel), which
+// are then merged together and passed into another HStar3 at depth 8 which
+// computes the overall tree root update.
+//
+// For that reason, Update doesn't invoke NodeStorage.Set for the topmost level
+// nodes. If it did then chained Updates would Set the borderline nodes twice.
+//
+// Warning: This call modifies the updates slice in-place, so the caller must
+// ensure to not reuse it.
+func (h HStar3) Update(updates []NodeUpdate, top uint, ns NodeStorage) ([]NodeUpdate, error) {
+	for d := h.depth; d > top; d-- {
+		var err error
+		if updates, err = h.updateAt(updates, d, ns); err != nil {
+			return nil, fmt.Errorf("depth %d: %v", d, err)
+		}
+	}
+	return updates, nil
+}
+
+// updateAt applies the given node updates at the specified tree level.
+// Returns the updates that propagated to the level above.
+func (h HStar3) updateAt(updates []NodeUpdate, depth uint, ns NodeStorage) ([]NodeUpdate, error) {
+	// Apply the updates.
+	for _, upd := range updates {
+		ns.Set(upd.ID, upd.Hash)
+	}
+	// Calculate the updates that propagate to one level above.
+	newLen := 0
+	for i, ln := 0, len(updates); i < ln; i, newLen = i+1, newLen+1 {
+		sib := updates[i].ID.Sibling()
+		left := updates[i].Hash
+		var right []byte
+		if next := i + 1; next < ln && updates[next].ID == sib {
+			// The sibling is the right child here, as updates are sorted.
+			right = updates[next].Hash
+			i = next // Skip the next update in the outer loop.
+		} else {
+			// The sibling is not updated, so fetch the original from NodeStorage.
+			var err error
+			if right, err = ns.Get(sib); err != nil {
+				return nil, err
+			}
+			if isLeftChild(sib) {
+				left, right = right, left
+			}
+		}
+		hash := h.hash(left, right)
+		updates[newLen] = NodeUpdate{ID: sib.Prefix(depth - 1), Hash: hash}
```

Comment:
For me, it would be clearer if newLen was incremented here instead of in the loop post section.